### ISSUE TITLE
GraphQL: Third party input type not registered properly

### ISSUE THREAD

**Issue by DanielHaring** [craftcms/cms#7551]
### Description

When a plugin registers a _GraphQL_ type through `RegisterGqlTypesEvent`, fields of this type correctly can be used inside queries. However, when trying to save data through mutations, I’m getting a _"Tried to load an unregistered type"_ error.

This only occurs with commands passed directly to the _GraphQL_ endpoint (`Craft::$app->gql->executeQuery`). When firing the exact same command through _GraphiQL_, everything works fine.

I experienced this issue with _ether/simplemap_ while trying to save a location:

```graphql
mutation saveFoo($id: ID, $location: Ether_MapInput) {
    save_foo_bar_Entry(id: $id, location: $location) {
        id,
        on foo_bar_Entry {
            location {
                lat,
                lng
            }
        }
    }
}
```

This will lead to:

> Tried to load an unregistered type "Ether_MapInput".

I was able to trace this back to:

https://github.com/craftcms/cms/blob/2e6d061031891a37ecec257507051d5b450d9d53/src/services/Gql.php#L1288-L1293

While the type is registered correctly, the input type (`\ether\simplemap\integrations\graphql\MapType::getInputType`) is completely ignored.

I was able to fix this inside my `~/modules/Module.php`:

```php
TypeLoader::registerType(MapType::getName() . 'Input', MapType::class . '::getInputType');
```

However, I think Craft CMS should take care of this itself.

### Additional info

- Craft CMS version: 3.5.17.1
- PHP version: 7.3.21
- Database driver & version: MySQL 5.7.30
- Plugins & versions: ether/simplemap v3.9.0.2

**Comment by andris-sevcenko** [comment]
Are you using the right schema when using the query outside of GraphQL API? If there was an error in how the type is registered, it should be present in both scenarios, all other things being equal.

I'm unable to reproduce this with the Craft and Plugin versions you're using.

**Comment by DanielHaring** [comment]
Well, I use a custom (private) schema that is allowed to save the corresponding entry type (all other fields work fine).

I select it via `Craft::$app->gql->getSchemaByUid('foo')` when firing against the _GraphQL_ endpoint and via the dropdown inside _GraphiQL_.

Additionally, I noticed that when I specify the non-input type (`Ether_Map`) I don't get the error _"Tried to load an unregistered type"_:

```graphql
mutation saveFoo($id: ID, $location: Ether_Map) {
    save_foo_bar_Entry(id: $id, location: $location) {
        id
    }
}
```

The result is:

> Variable "$location" of type "Ether_Map" used in position expecting type „Ether_MapInput".

Which is completely correct. So to me it looks like the regular type (`Ether_Map`) is registered correctly, while the input type (`Ether_MapInput`) is not.

**Comment by andris-sevcenko** [comment]
So when you say "GraphQL endpoint" you mean the internal API, not an external endpoint, gotcha.

Can you check if the value of `prebuildSchema` here https://github.com/craftcms/cms/blob/c4b999b325736968a05d3bcc1a5a2444dfbe9323/src/services/Gql.php#L366 differs when you run the mutation in GraphiQL and if you use the internal API?

It would probably be useful to figure out what the difference is between those two requests.

**Comment by DanielHaring** [comment]
Sure!

There is indeed a difference:

GraphiQL: `(boolean) true`
Internal API: `(boolean) false`

**Comment by andris-sevcenko** [comment]
Actually, that doesn't mean anything. I thought it would be because the input type would be "invisible", but that is not the case - in order to resolve the mutation, it needs to be created on execution on Craft's end. The very act of creating it checks the arguments and loads the input types.

The only scenario I can think of off the top of my head is if the Bar entry type for the Foo section does not actually have the maps field defined, which is unlikely.

Is it possible to get a DB dump, composer.* files _and_ the custom code you're running? If so, can you email them to support@craftmcs.com and reference this issue?

**Comment by andris-sevcenko** [comment]
> When firing the exact same command through GraphiQL, everything works fine.

This was a little misleading. It _does_ work if you fire the same exact command and don't use any variables, as opposed to the code implementation. If you execute the mutation in question and use variables in GraphiQL, it does not work and breaks with the same error.

If you use any complex type in variables, GraphQLhas to figure out what that type is and it loads this type directly. If, however, you just pass in the arguments without using variables, GraphQL will just figure out what the type was from the context, which in this case was the mutation that had all the type information.

----

So, in short - the problem occurred here when the Ether_MapInput type was loaded directly.

To fix this, Ether should update the plugin ([here](https://github.com/ethercreative/simplemap/blame/v3/src/integrations/graphql/MapType.php#L104-L194)) to use the type name instead of the class name, when creating the types, otherwise, it's impossible to look them up.

Granted, we could do a better job on the documentation front of this.
